sw/device/lib/crypto/drivers/entropy.c

```diff
@@ -0,0 +1,183 @@
+  }
+  return OK_STATUS();
+}
+
+status_t entropy_csrng_instantiate(
+    hardened_bool_t disable_trng_input,
+    const entropy_seed_material_t *seed_material) {
+  return csrng_send_app_cmd(kBaseCsrng + CSRNG_CMD_REQ_REG_OFFSET,
+                            (entropy_csrng_cmd_t){
+                                .id = kEntropyDrbgOpInstantiate,
+                                .disable_trng_input = disable_trng_input,
+                                .seed_material = seed_material,
+                                .generate_len = 0,
+                            });
+}
+
+status_t entropy_csrng_reseed(hardened_bool_t disable_trng_input,
+                              const entropy_seed_material_t *seed_material) {
+  return csrng_send_app_cmd(kBaseCsrng + CSRNG_CMD_REQ_REG_OFFSET,
+                            (entropy_csrng_cmd_t){
+                                .id = kEntropyDrbgOpReseed,
+                                .disable_trng_input = disable_trng_input,
+                                .seed_material = seed_material,
+                                .generate_len = 0,
+                            });
+}
+
+status_t entropy_csrng_update(const entropy_seed_material_t *seed_material) {
+  return csrng_send_app_cmd(kBaseCsrng + CSRNG_CMD_REQ_REG_OFFSET,
+                            (entropy_csrng_cmd_t){
+                                .id = kEntropyDrbgOpUpdate,
+                                .seed_material = seed_material,
+                                .generate_len = 0,
+                            });
+}
+
+status_t entropy_csrng_generate_start(
+    const entropy_seed_material_t *seed_material, size_t len) {
+  // Round up the number of 128bit blocks. Aligning with respect to uint32_t.
+  // TODO(#6112): Consider using a canonical reference for alignment operations.
+  const uint32_t num_128bit_blocks = (len + 3) / 4;
+  return csrng_send_app_cmd(kBaseCsrng + CSRNG_CMD_REQ_REG_OFFSET,
+                            (entropy_csrng_cmd_t){
+                                .id = kEntropyDrbgOpGenerate,
+                                .seed_material = seed_material,
+                                .generate_len = num_128bit_blocks,
+                            });
+}
+
+status_t entropy_csrng_generate_data_get(uint32_t *buf, size_t len) {
+  for (size_t i = 0; i < len; ++i) {
+    // Block until there is more data available in the genbits buffer. CSRNG
+    // generates data in 128bit chunks.
+    if (i % kEntropyCsrngBitsBufferSize == 0) {
+      uint32_t reg;
+      do {
+        reg = abs_mmio_read32(kBaseCsrng + CSRNG_GENBITS_VLD_REG_OFFSET);
+      } while (!bitfield_bit32_read(reg, CSRNG_GENBITS_VLD_GENBITS_VLD_BIT));
+    }
+    buf[i] = abs_mmio_read32(kBaseCsrng + CSRNG_GENBITS_REG_OFFSET);
+  }
+  return OK_STATUS();
+}
+
+status_t entropy_csrng_generate(const entropy_seed_material_t *seed_material,
+                                uint32_t *buf, size_t len) {
+  TRY(entropy_csrng_generate_start(seed_material, len));
```

Comment:
Extending on my earlier comment: we could define sthg like `CRYPTO_RETURN_IF_ERROR` and use it here.

Reply:
Let's close on the cryptolib error handling details here https://github.com/lowRISC/opentitan/issues/14549. We can update the drivers after we have more details documented.